components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/internal/service/impl/password/PasswordRecoveryManagerImpl.java

```diff
@@ -202,6 +206,70 @@ public PasswordResetCodeDTO confirm(String confirmationCode, String tenantDomain
         return buildPasswordResetCodeDTO(confirmationCode);
     }
 
+    /**
+     * Validate the code given for password recovery and return the password reset code.
+     *
+     * @param otp              One Time Password
+     * @param confirmationCode Confirmation code
+     * @param tenantDomain     Tenant domain
+     * @param properties       Meta properties in the confirmation request
+     * @return PasswordResetCodeDTO {@link PasswordResetCodeDTO} object which contains password reset code
+     * @throws IdentityRecoveryException Error while confirming password recovery
+     */
+    @Override
+    public PasswordResetCodeDTO confirm(String otp, String confirmationCode, String tenantDomain,
+                                        Map<String, String> properties) throws IdentityRecoveryException {
+
+        validateTenantDomain(tenantDomain);
+        UserAccountRecoveryManager userAccountRecoveryManager = UserAccountRecoveryManager.getInstance();
+        // In the recovery scenarios which are not OTP based, the confirmation code is a combination of the
+        // recovery flow id and another UUID generated. Hence, we need to get the recovery flow id from the
+        // confirmation code. In the OTP based recovery scenario, the confirmation code is the recovery flow id.
+        String[] ids = confirmationCode.split("\\.");
+        String recoveryFlowId;
+        String code;
+        if (ids.length == 2) {
+            recoveryFlowId = ids[0];
+            code = confirmationCode;
+        } else {
+            recoveryFlowId = confirmationCode;
+            code = otp;
+        }
+        // Get Recovery data.
+        UserRecoveryData userRecoveryData = userAccountRecoveryManager
+                .getUserRecoveryDataFromFlowId(recoveryFlowId, RecoverySteps.UPDATE_PASSWORD);
+        int attempt = userRecoveryData.getAttempt();
+        if (!tenantDomain.equals(userRecoveryData.getUser().getTenantDomain())) {
+            throw Utils.handleClientException(
+                    IdentityRecoveryConstants.ErrorMessages.ERROR_CODE_USER_TENANT_DOMAIN_MISS_MATCH_WITH_CONTEXT,
+                    tenantDomain);
+        }
+        String domainQualifiedName = IdentityUtil.addDomainToName(userRecoveryData.getUser().getUserName(),
+                userRecoveryData.getUser().getUserStoreDomain());
+        if (StringUtils.equals(code, userRecoveryData.getSecret())) {
+            if (log.isDebugEnabled()) {
+                log.debug("Valid confirmation code for user: " + domainQualifiedName);
+            }
+            return buildPasswordResetCodeDTO(code);
+        } else {
+            if ((attempt + 1) >= Integer.parseInt(Utils.getRecoveryConfigs(IdentityRecoveryConstants.ConnectorConfig.
```

Comment:
Since this is tenant level config, we need to pass the tenant domain as well. And since the config is used only once in the method, it'll be redundant to use a variable for now.

Updated `attempt` to `failedAttempts`.